### PR TITLE
Add stream of streams retrieval method to the event store repo

### PR DESCRIPTION
**Pull request by jaceko**

### CODE REVIEW

File: event-sourcing/event-repository/event-repository-core/src/main/java/uk/gov/justice/services/eventsourcing/repository/core/EventRepository.java [hunk 1]
```diff
@@ -59,4 +59,9 @@
      */
     Long getCurrentSequenceIdForStream(final UUID streamId);
 
+    /**
+     * Returns stream of envelope streams. Envelopes in the nested stream are ordered by sequenceId
+     * @return the stream of envelope streams
+     */
+    Stream<Stream<JsonEnvelope>> getStreamOfEventStreams();
```
Comment (matt-rich) [inline event-sourcing/event-repository/event-repository-core/src/main/java/uk/gov/justice/services/eventsourcing/repository/core/EventRepository.java]: Is it slightly clearer to have 'All' in the naming somewhere, to show that it is retrieving all events in stream of stream format.

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java [hunk 2]
```diff
@@ -95,6 +102,79 @@ public void shouldGetAll() throws Exception {
...
+    @Test
+    public void shouldGetStreamOfStreams() throws Exception {
+        final UUID streamId1 = UUID.fromString("4b4e80a0-76f7-476c-b75b-527e38fb251e");
+        final UUID streamId2 = UUID.fromString("4b4e80a0-76f7-476c-b75b-527e38fb252e");
+        final UUID streamId3 = UUID.fromString("4b4e80a0-76f7-476c-b75b-527e38fb253e");
+
+        final EventLog event1 = new EventLog(null, streamId1, null, null, null, null, null);
+        final EventLog event2 = new EventLog(null, streamId2, null, null, null, null, null);
+        final EventLog event3 = new EventLog(null, streamId3, null, null, null, null, null);
+        final JsonEnvelope envelope1 = DefaultJsonEnvelope.envelope().build();
```
Comment (matt-rich) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java]: static import DefaultJsonEnvelope ?

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java [hunk 3]
```diff
@@ -95,6 +102,79 @@ public void shouldGetAll() throws Exception {
         verify(logger).trace("Retrieving all events");
     }
 
+    @Test
+    public void shouldGetStreamOfStreams() throws Exception {
+        final UUID streamId1 = UUID.fromString("4b4e80a0-76f7-476c-b75b-527e38fb251e");
+        final UUID streamId2 = UUID.fromString("4b4e80a0-76f7-476c-b75b-527e38fb252e");
+        final UUID streamId3 = UUID.fromString("4b4e80a0-76f7-476c-b75b-527e38fb253e");
+
+        final EventLog event1 = new EventLog(null, streamId1, null, null, null, null, null);
```
Comment (matt-rich) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java]: Could we move some of this into a private creator as it is reused ?
eventLogOf(steadId1) etc ? - similar to below test.

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java [hunk 4]
```diff
@@ -103,6 +104,28 @@ public void shouldReturnAllEventsOrderedBySequenceId() throws InvalidSequenceIdE
 
     }
 
+    @Test
+    public void shouldReturnStremOfStreamIds() throws Exception {
```
Comment (matt-rich) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java]: typo on Strem

File: persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java [hunk 5]
```diff
@@ -42,32 +43,51 @@ protected DataSource getDataSource() {
...
 
     protected Stream<T> streamOf(final PreparedStatementWrapper psWrapper) throws SQLException {
 
         final ResultSet resultSet = psWrapper.executeQuery();
 
-        return StreamSupport.stream(new Spliterators.AbstractSpliterator<T>(
+        return streamOf(psWrapper, resultSet);
+    }
+
+    protected <U> Stream<U> streamOf(final PreparedStatementWrapper psWrapper, final ResultSet resultSet, final Function<ResultSet, U> rs2EntityFunction) {
```
Comment (matt-rich) [inline persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java]: Is there a better name for this ? rs2EntityFunction
Is it resultSetEntityMapperFunction or something?

Might be useful to have a javadoc comment, explaining what U is etc.

File: persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java [hunk 6]
```diff
@@ -42,32 +43,51 @@ protected DataSource getDataSource() {
...
+        return StreamSupport.stream(new Spliterators.AbstractSpliterator<U>(
                 Long.MAX_VALUE, Spliterator.ORDERED) {
```
Comment (matt-rich) [inline persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java]: Can we static import these ?

File: event-sourcing/event-repository/event-repository-core/src/main/java/uk/gov/justice/services/eventsourcing/repository/core/EventRepository.java (hunk 1 above)
Comment (jaceko) [inline event-sourcing/event-repository/event-repository-core/src/main/java/uk/gov/justice/services/eventsourcing/repository/core/EventRepository.java]: done

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java (hunk 3 above)
Comment (jaceko) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java]: done

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java (hunk 4 above)
Comment (jaceko) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java]: fixed

File: persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java (hunk 5 above)
Comment (jaceko) [inline persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java]: done

File: persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java (hunk 6 above)
Comment (jaceko) [inline persistence/persistence-jdbc/src/main/java/uk/gov/justice/services/jdbc/persistence/AbstractJdbcRepository.java]: we could but isn't Spliterator.ORDERED more clear than just ORDERED?
and Long.MAX_VALUE than just MAX_VALUE

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java [hunk 7]
```diff
@@ -219,4 +219,8 @@ public void shouldReturnTestSequenceId() {
         assertThat(jdbcEventRepository.getCurrentSequenceIdForStream(STREAM_ID), equalTo(VERSION_1));
     }
 
+    private EventLog eventLogOf(final UUID streamId1) {
```
Comment (matt-rich) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/JdbcEventRepositoryTest.java]: should be steamId but streamId1

**Pull request merged**